Partial fix of bcf memory leak

sem-geologist opened the pull request:
### Description of the changes:
- unify cythonish and pure pythonic parsing function interfaces
- exclude dask from wrapping the numpy array with `.compute()` method if `lazy=False`

### Detailed description:
To begin with, #1773 `dask` array bug makes the library of little use at this moment, compared with pre-v1.2 hyperspy versions.
The bug could be partly neglected by excluding `dask` involvement when `lazy=False`.
Inspection of the code made me realize that dask/lazy loading loop was implemented in current way due to complicated bcf library codebase.
The bcf code included in hyperspy were (still is) planned to be used with some other custom software of mine,  where some functions in hyperspy are stripped and the code in some parts looks weird (e.g. class with only `__init__` method where more proper implementation would be  a dictionary.
The `if lazy` part is complicated too, as pure python and cython parsing functions have different interfaces. The pure python parsing method was written explicitly for hyperspy and differs from cython implementation where it is a function. Furthermore some classes are pointless in bcf for hyperspy, as io anyway accepts only dictionary. Blown-up instances of classes particularly are dangerous with dask delayed. So...

### Progress of the PR
- [x] get rid of HyperMap class, ~~same can be achieved with `dict`~~; better - don't do persistant parsing, get rid of that method, don't create abstract atributes/ dicts, define relevant stuff straight into final dict.
- [x] move the `.py_parse_hypermap` method out of class, refactor it into function
- [x] unify parsing function interfaces (python/cython), relocate shape, dtype generation code from those functions into single code. 
- [x] refactor the `if lazy` logic, which will get more simple with simple interface
- [x] check tests,
- [x] ready for review.

sem-geologist commented:
@francisco-dlp , maybe this could go to next_patch?

francisco-dlp commented:
No, it cannot because you have branched out for  `next_minor`. However, you could fix that and resubmit it to `next_patch`.

francisco-dlp commented on hyperspy/io_plugins/bcf.py:
```diff
@@ -946,246 +908,217 @@ def parse_hypermap(self, index=None,
-            cutoff_chan = None
-
+            max_chan = self.header.estimate_map_channels(index=index)
+        shape = (-(-self.header.image.height // downsample),
```
Why are you using the negative sign in `(-(-`? If it is to avoid getting a `0` I would suggest the more explicit:

```python


shape = (self.header.image.height // downsample, self.header.image.width // downsample, max_chan)
shape = [ss if ss >= 1 else 1 for ss in shape]
```

sem-geologist replied:
there is such math function: floor, what I need here is roof, that is the trick

sem-geologist replied:
i.e. `5 // 2` would give 2, but I need 3, but I also need that `4 // 2` would give 2 instead of returning 2 in both cases.

sem-geologist replied:
I could add comment that it is roof, maybe it would be clear then what `-(-` does.

francisco-dlp replied:
Then, maybe what you need is ``np.ceil`` or ``math.ceil``

sem-geologist replied:
I think math.ceil is more approprate as it returns integer. I wonder how did I missed that simple solution...

francisco-dlp commented on hyperspy/io_plugins/bcf.py:
```diff
@@ -946,246 +908,217 @@ def parse_hypermap(self, index=None,
+            x_pix, chan1, chan2, dummy1, flag, dummy_size1, n_of_pulses,\
+                data_size2 = strct_unp('<IHHIHHHI',
+                                       buffer1[offset:offset + 22])
+            pix_idx = (x_pix // dwn_factor) + ((-(-width // dwn_factor)) *
```
As above

francisco-dlp commented on hyperspy/io_plugins/bcf.py:
```diff
@@ -946,246 +908,217 @@ def parse_hypermap(self, index=None,
+            else:
+                vfa[max_chan * pix_idx:chan1 + max_chan * pix_idx] +=\
+                    pixel[:chan1]
+    vfa.resize((-(-height // dwn_factor),
```
As above

francisco-dlp reviewed (commented):
Other than the small style comment, it looks good to me.

sem-geologist commented:
aaa I dont know how it is doing that, but it is pulling some stuff from next_minor. I need to find how to drop those files...

sem-geologist commented:
also I see in tests that some numpy deprecation appeared (`fromstring` deprecated, it should be replaced with `frombuffer`)

francisco-dlp commented:
Regarding ``fromstring``, @ericpre fixed it in #1799 & #1806

sem-geologist committed:
rm: upstream with next
removal of lxml dependency

sem-geologist committed:
bcf-refactoring

sem-geologist committed:
Merge branch 'RELEASE_next_patch' into bcf-refactor-1

sem-geologist committed:
convert all numpy fromstring

into frombuffer

sem-geologist committed:
not all numpy

needs to do frombuffer

sem-geologist committed:
some frombuffer

are imutable;
changed into fromiter

sem-geologist committed:
try_to_fix

sem-geologist committed:
revert back to numpy.fromstring

as from buffer is immutable, and fromiter introduces
parsing errors

sem-geologist committed:
switch one array reading

into frombuffer

sem-geologist committed:
now its really ugly numpy

sem-geologist commented:
I can't understand how #1806 could pass, as replacing `fromstring` into `frombuffer` is not enough. thanks to new numpy's way in too-fast, non-discussed rushed changes. The main weakness of this deprecation is that frombuffer is not direct replacement of `fromstring` as first returns immutable array where e.g. byteswich() is then not working, while the other returns a new array. The direct replacement should be like this: `np.frombuffer(a, *args).copy()`. so i added `.copy()` where appropriate to comply with newer numpy and pass all tests.

@francisco-dlp , the PR was reuploaded removing traces of next-major. I replaced all `-(-a//b)` constructions into `ceil(a / b)` (using `math.ceil`). I would like this to be merged, so I could start to clean up way for #1801 .  I would like to have it separate from this PR.
